src/Core/ConstraintValidator/ZipCodeRangeValidator.php

```diff
@@ -0,0 +1,123 @@
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\ConstraintValidator;
+
+use PrestaShop\PrestaShop\Core\CommandBus\CommandBusInterface;
+use PrestaShop\PrestaShop\Core\Domain\Country\Exception\CountryConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Country\Query\GetCountryZipCodeRequirements;
+use PrestaShop\PrestaShop\Core\Domain\Country\QueryResult\CountryZipCodeRequirements;
+use PrestaShop\PrestaShop\Core\Form\FormChoiceProviderInterface;
+use Symfony\Component\Translation\TranslatorInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+
+/**
+ * Validate zip codes in range
+ */
+class ZipCodeRangeValidator extends ConstraintValidator
+{
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var CommandBusInterface
+     */
+    private $queryBus;
+
+    /**
+     * @var FormChoiceProviderInterface
+     */
+    private $countryChoiceProvider;
+
+    /**
+     * @param TranslatorInterface $translator
+     * @param CommandBusInterface $queryBus
+     * @param FormChoiceProviderInterface $countryChoiceProvider
+     */
+    public function __construct(
+        TranslatorInterface $translator,
+        CommandBusInterface $queryBus,
+        FormChoiceProviderInterface $countryChoiceProvider
+    ) {
+        $this->translator = $translator;
+        $this->queryBus = $queryBus;
+        $this->countryChoiceProvider = $countryChoiceProvider;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws CountryConstraintException
+     */
+    public function validate($value, Constraint $constraint)
+    {
+        $countryId = (int) $value['country'];
+        $selectedCountries = [$countryId];
+        $zipCode = $value['zipCode'];
+
+        if (null === $zipCode) {
+            return;
+        }
+
+        if ($countryId === 0) {
+            $countries = $this->countryChoiceProvider->getChoices();
+            $selectedCountries = array_map('intval', $countries);
+        }
+
+        foreach ($selectedCountries as $country) {
+            /** @var CountryZipCodeRequirements $requirements */
+            $requirements = $this->queryBus->handle(new GetCountryZipCodeRequirements($country));
```

Comment:
Yes indeed
If a Handler needs to use another Handler, we can inject it directly 🤔 

Command Bus must act as a border between legacy code and modern code.
Every code on the part that dispatches Queries and Commands must be modern: no usage of legacy classes, decoupled, testable, follow SOLID code
On the handler side we can use Adapter that rely on legacy code.

Reply:
Well im not sure what to do then. If im not wrong this query will be used in https://github.com/PrestaShop/PrestaShop/pull/15300 PR. Also in symfony validator. So instead of query there should be an adaptor service with core interface created and injected?

Reply:
Yes, let's go for an adapter service with a core interface.
This adapter service can be injected in the places where it's needed. If it is needed on the domain side, by a handler, then it's injected in the handler. If it's needed by something else that is not a handler, then it's injected too.

Reply:
So in this case instead of query handler i will have this new service. But what about query result? What namespace should it have if we remove queryBus out of equation?